hub/core/chunk_engine/tests/test_chunk_generator.py

```diff
@@ -0,0 +1,171 @@
+from hub.core.chunk_engine.generator import chunk
+
+from util import (
+    make_dummy_byte_array,
+    get_random_chunk_size,
+    get_random_num_samples,
+    get_random_partial,
+    assert_valid_chunk,
+)
+
+
+TRIALS_PER_TEST = 10
+
+
+def test_perfect_fit():
+    """This test is for the edge case where: (num_bytes % chunk_size == 0). In other words, bytes fit perfectly into chunks with none left over."""
+
+    for _ in range(TRIALS_PER_TEST):
+        chunk_size = get_random_chunk_size()
+        num_samples = get_random_num_samples()
+
+        content = make_dummy_byte_array(chunk_size * num_samples)
+        gen = chunk(content, previous_num_bytes=None, chunk_size=chunk_size)
+
+        piece_count = 0
+        for piece, relative_chunk_index in gen:
+            assert_valid_chunk(piece, chunk_size)
+
+            expected_start = piece_count * chunk_size
```

Comment:
This is the age-old problem of duplicating your code into the tests. The only thing this gives us is twice as much code to change for any modification. 

Tests should be much simpler than this. Take the examples you give in the docstring, and test them. Then maybe a few more for some other corner cases, but that's it. If we over-engineer the tests we'll just end up right where we started.

Reply:
you're absolutely right